src/components/heading/heading.scss

```diff
@@ -0,0 +1,125 @@
+@use '../../design-tokens/colors.yml';
+@use '../../design-tokens/layout.yml';
+@use "../../design-tokens/motion.yml";
+@use '../../design-tokens/opacity.yml';
+@use '../../design-tokens/sizes.yml';
+@use '../../mixins/focus';
+@use '../../mixins/headings';
+@use '../../mixins/ms';
+@use '../../mixins/unit';
+@use 'sass:math';
+
+$permalink-icon-size: sizes.$control-icon-size;
+$permalink-radius: sizes.$radius-full;
+$permalink-size: sizes.$control-height;
+
+$min-width-permalink-shift: layout.$max-width-prose + $permalink-size * 4;
+$max-width-permalink-shift: ($min-width-permalink-shift * 16 - 1) / 16;
+
+/**
+ * This containing element may be applied to a heading directly to apply a
+ * heading style, or to a containing element if you want to display a heading
+ * and permalink.
+ *
+ * 1. We default to the lowest heading level supported by HTML5, which forces
+ *    the template designer to specify an intended visual heading level.
+ * 2. Sets up the optional permalink to position itself relative to the
+ *    containing element.
+ */
+
+.c-heading {
+  @include headings.level(6); /* 1 */
+  position: relative; /* 2 */
+}
+
+/**
+ * Level modifiers
+ *
+ * These will apply the styles for a specific heading level.
+ *
+ * We start at `-2` so template designers have three sizes larger than a default
+ * `<h1>` to apply thoughtfully within templates.
+ *
+ * We end at 3 because that is the maximum depth we support with distinct
+ * visual differences.
+ */
+
+@for $level from -2 through headings.$max-level {
+  $level-segment: ms.step-class-segment($level);
+
+  .c-heading--level-#{$level-segment} {
+    @include headings.level($level);
+  }
+}
+
+/**
+ * If `c-heading` is applied to a containing element, this class must be applied
+ * to the inner heading to avoid unintended font sizes.
+ */
+
+.c-heading__element {
+  font: inherit;
+
+  /**
+   * Prevent heading from overlapping permalink until the viewport is wide
+   * enough for the permalink to position itself outside the container.
+   */
+
+  @media (max-width: $max-width-permalink-shift) {
+    padding-right: unit.swap($permalink-size, rem);
+  }
+}
+
+/**
+ * Optional permalink to be included adjacent to the heading itself.
+ *
+ * 1. We define the element's `font-size` using `rem` so its size will remain
+ *    consistent between differently sized headings in the same article.
+ * 2. The link's shape is not visible unless focused, so applying some negative
+ *    offset based on the inner padding makes the link appear better aligned
+ *    with content above and below.
+ *
+ * @see http://codepen.io/marcysutton/pen/rLKvgZ
+ */
+
+.c-heading__permalink {
```

Comment:
I'm open to a different name, but I think adding `-icon` is confusing, because the icon is nested _within_ this element. There's also no reason you couldn't use a glyph like `#`, which isn't really an icon.

Reply:
That's fair. I'm not sure what a better name would be, but here are a few thoughts and questions that may inspire some brainstorming:

1. Does the pattern break if content longer than a single icon or glyph is used?
2. Does the twig template support using something other than the link icon?
3. Does the pattern name clearly communicate that it's not either of these?
    - `<h1><a>Text</a></h1>`
    - `<a><h1>Text</h1></a>`
4. Does `permalink` provide more context than just `link`?

If we don't have a better idea I'm fine shipping `permalink` but thought it might be helpful to discuss further.

Reply:
> Does the pattern break if content longer than a single icon or glyph is used?

Yes!

> Does the twig template support using something other than the link icon?

No, but it easily could.

> Does the pattern name clearly communicate that it's not either of these? 

No, but I don't know if it's always necessary for a pattern name to clealry communicate what it _cannot_ do. It's a nice bonus, but I think the only suggestion so far (`link-icon`) sacrifices what the pattern _is_ in favor of what the pattern _isn't_. And I can't think of any alternatives.

> Does `permalink` provide more context than just `link`?

I think so, since the heading element itself can contain a link (`<h2>Hello <a href="...">world</a></h2>`). I wanted it to be clear this wasn't for _any_ link, but specifically the permalink.

Reply:
Yeah that makes sense. Cool, I'm fine with permalink then. Thanks!